### ISSUE TITLE
Support installing .zip files

### ISSUE THREAD

**Issue by denelon** [microsoft/winget-cli#140]
I want to be able to have winget install apps inside of .zip files.

Related to:
* https://github.com/microsoft/winget-cli-restsource/issues/117

# Proposed technical implementation details (optional)

**Comment by utybo** [comment]
This would be great! Gradle has always required a package manager on Windows (I'm currently using it with Scoop) and having it on winget would be fantastic, but it's a .zip file.

Also, wouldn't zip files (and standalone .exe apps) require some more complicated setup steps that would have to be described in winget's YAML file? e.g. adding things to PATH, creating a desktop link, creating a Start menu entry etc.

**Comment by zuozishi** [comment]
>

> This would be great! Gradle has always required a package manager on Windows (I'm currently using it with Scoop) and having it on winget would be fantastic, but it's a .zip file.
> 
> Also, wouldn't zip files (and standalone .exe apps) require some more complicated setup steps that would have to be described in winget's YAML file? e.g. adding things to PATH, creating a desktop link, creating a Start menu entry etc.

And environment variable.

**Comment by DennisGaida** [comment]
I think there should be a default _Apps_ folder maybe within the %USERPROFILE%. Unzipped zips should be per default copied there maybe within a subfolder. The same may also happen for PowerShell Scripts (PS1) and EXEs.

**Comment by PedroLamas** [comment]
I was about to create a manifest for Paint.net when I noticed that .zip is not supported... yet! 🙂

**Comment by pashagolub** [comment]
A lot of .msi installers are packed into .zip files because you cannot execute .msi downloaded from internet. ;-) 

I wanted to add PostgreSQL support, but lack of zip extraction is a problem here.

**Comment by maxandersen** [comment]
related to https://github.com/microsoft/winget-pkgs/issues/580

Right now Winget can't be used to package/install anything without complex installers.

**Comment by chausner** [comment]
I think there are actually two issues or requests related to ZIP:
1. Support software that is provided as ZIP archives for plain xcopy deployment
2. Support installers (MSI, EXE, etc.) that are made available for download in a ZIP file (e.g. https://mh-nexus.de/de/downloads.php?product=HxD20)

**Comment by timsneath** [comment]
@aclinick, ideally it would be a) unzip a ZIP file; b) (optionally) run a batch file script to perform some configuration. Would be reasonable if there's also an uninstall script that reverses the latter. 

We'd use this for installing Dart and Flutter on Windows.

**Comment by marypcbuk** [comment]
I'd suggest ZIP support with a parameter to install to either Program Files or a specified location given in the command

**Comment by shayvaturi** [comment]
If we have a solution that works for everybody (Scoop), why trying to duplicate it?

**Comment by KevinMarquette** [comment]
> @aclinick, ideally it would be a) unzip a ZIP file; b) (optionally) run a batch file script to perform some configuration. Would be reasonable if there's also an uninstall script that reverses the latter.

Having your PowerShell scripts in the zip that to be executed opens up a lot of options to universally install anything. PowerShell can be used to fill any gaps that WinGet has not covered yet. Every one of my custom Chocolatey packages is basically a folder full of files with an install.ps1.

**Comment by tap52384** [comment]
Adobe Creative Cloud is another example of an installer inside a zip file:
<https://ccmdl.adobe.com/AdobeProducts/KCCC/CCD/5_1/win64/ACCCx5_1_0_407.zip>

This seems like the installer that supports silent installation, not the smaller one that requires an internet connection.

It seems that "tar.exe" is a part of Windows 10, so this could be really useful.

**Comment by bitcrazed** [comment]
@tap52384 😲

**Comment by benc-uk** [comment]
I think we need to support installing standalone binary only tools like `kubectl` where you typically just download a exe and put into your path. That leads to the question where in path to put it?

This is how a lot of modern developer tools are distributed

**Comment by ElSrJuez** [comment]
Or... post an issue on @Sysinternals to publish their packages as MSIX?

**Comment by WtfJoke** [comment]
> Or... post an issue on @Sysinternals to publish their packages as MSIX?

There are ton of tools which doesnt need an installer, sysinternals is just one of them.

You download an exe add it to path and thats it. I usually preferr them, no hastle at all. And should be easy to implement for winget too. 

Curl, minkube, github cli, putty, or all sysinternals tool as you mentioned. Or any go executable/programm... even adoptopenjdk would offer a api for downloading zip.

**Comment by shayvaturi** [comment]
> > Or... post an issue on @Sysinternals to publish their packages as MSIX?
> 
> There are ton of tools which doesnt need an installer, sysinternals is just one of them.
> 
> You download an exe add it to path and thats it. I usually preferr them, no hastle at all. And should be easy to implement for winget too.
> 
> Curl, minkube, github cli, putty, or all sysinternals tool as you mentioned. Or any go executable/programm... even adoptopenjdk would offer a api for downloading zip.

That's exactly what you have in scoop - extracts the zip file and creates a shim to the exe. Shims folder is in the path.

We can use scoop and winget in parallel, or take the functionality from scoop and add it to winget if we want the whole solution in one place.

**Comment by UzEE** [comment]
+1 for adding the ability to execute a `.ps1` or `.bat` script after extracting a ZIP file. Ideally, this should be a file that may not be included in the original package. This should allow us to work with most types of packages distributed as ZIP files.

For example, [MSI Afterburner](https://www.msi.com/page/afterburner) setup is distributed in a ZIP file. Ideally, we should be able to include the script along with the manifest in winget-pkgs as well.

**Comment by ooeygui** [comment]
The Microsoft Edge Robotics would also like to see this implemented for the distribution of the [Robot Operating System (ROS) on Windows](http://aka.ms/ros) and 3rd party ROS packages. 

To enable migration to Winget from other package managers, zip file deployment and powershell scripts is critical.

**Comment by bpbradley** [comment]
I will add a +1 that this feature is required for useful deployment of many cross platform toolchains. Distributing binaries on github release pages is fairly common practice.

**Comment by chrpai** [comment]
IMO Sysinternals could use an installer so I created one.  

1. Prompt user for location to download zip from (self hosting offline scenarios)  (Default Microsoft URL for silent/winget installation)

2. Download and extract ZIP to installation directory

3. Place installation directory at end of system path

4. Create start menu shortcut that points to installation directory.

UX is vastly improved. PR submitted:

https://github.com/microsoft/winget-pkgs/pull/1288

**Comment by chrpai** [comment]
@WtfJoke what's confusing?  This is basically how most choco packages do it.  They have a ools\chocolateyinstall.ps1 that downloads content from the vendor and then runs the installation commands.  Their nuspec file doesn't typically have the actual content because of legal redistribution rights issues.

**Comment by WtfJoke** [comment]
Confusion smiley is maybe not the right one. But In my opinion its not a solution to build for every tool an installer, because of a lacking feature of winget itself. 

So rather than patchworking for various software we should get it fixed/enhanced.

However I appreciate the work you put into it. Personally I wouldnt use it, since an unofficial installer is less trustworthy for me. Hope you understand

**Comment by chrpai** [comment]
I work in EUC for a fortune 30 company. I oversee the deployment of thousands of applications to 180,000 windows devices.  My life would be A LOT easier if every tool had an installer.  Tools such as WinGet, SCCM and Intune have application models.  When the app fits the model it just works.  When apps doing anything and everything they want it becomes impossible to create automation to support everything.  I hope you understand too.

**Comment by majkinetor** [comment]
After crating hundreeds of packages for various platforms and recently mostly working Chocolatey packages where I develop and maintain [coreteamrepo](https://gist.github.com/choco-bot/a14b1e5bfaf70839b338eb1ab7f8226f) I must say that PowerShell scripting within package should get TOP priority. Its simply not possible to do anything more complex relying only on yaml declarations and any "native" method (like ZIP extraction) will only reinvent the weel in a bad way.

**Comment by chrpai** [comment]
Zip files aren't packages or installers.  They are just archives that have no additional meta / context or standardized entry point for business rules / setup logic.  This feature request should be a non starter.

**Comment by doctordns** [comment]
A .ZIP installer would be a good addition. 

If you look at how Instal-PowerShell.ps1 script, used to install PowerShell 7 works. It enables a simple zip "installer" which is really just a find/download the zip file then expand it into the specified folder. Shouldn't Winget do something similar - if the manifest says zip, then just download it and expand it and let the user go from there.  

And of course, if there was some way to indicate that winget should create desktop/taskbar shortcuts to the relevant .exe in that expanded folder, double sweet.

**Comment by marypcbuk** [comment]
the problem with saying that Zip files should be out of scope because they aren't packaged installers is that would be the package manager trying to dictate the distribution format or hundreds of thousands of existing software tools, when Zip has been used as a distribution format for as many decades as I've been in the industry; to be useful, a Windows package manager needs to support the enormous variety of software out there that is one of the reasons Windows is so enduring an OS. 


> Zip files aren't packages or installers.  They are just archives that have no additional meta / context or standardized entry point for business rules / setup logic.  This feature request should be a non starter.

**Comment by chrpai** [comment]
Platforms have always set the standards for apps that run on then.  The problem is enforcement has been weak.   ISVs can and do put just about anything into a .ZIP and call it done.  Then I read comments from people that want the pkgmgr to do stuff with it like.

1) set destination directory
2) put application directory path or not
3) create shortcuts (path / name ) or not...

So just how are you supposed to do that programatically with no standard zip file format and no manifest / meta?

**Comment by majkinetor** [comment]
> Then I read comments from people that want the pkgmgr to do stuff with it like.

Yes, package script should be able to implement its own installer, if that is what one desires. I have done so in countless occassions. Just becuase vendor zip it, put a huge web page with instructions and call it a day doesn't mean we should follow up.

For example:

- [Gitlab-runner](https://github.com/majkinetor/au-packages/tree/master/gitlab-runner#package-parameters) - I implemented service option, autologon option and installdir, since for some of them u need external tools which is easy for package to depend on but may be hard for you to do manually. 
- [Rundeck](https://github.com/majkinetor/au-packages/tree/master/rundeck) - I implemented 10 options everybody needs but they really didn't want to put in easy to use manner for reasons I don't want to get into.
- [Signal](https://github.com/majkinetor/au-packages/tree/master/signal#package-parameters) - I added options people need but were not implemented upstream without finding specific issues on the pile (package installs oposite of those options by default, as that is what was asked by everbody)

Those are just couple of tools that are extensivelly used and didn't provide one line config. Its typical for jar applications to not provide any install options beside puting it in zip archive, so in order to be useful, you need to do so - you can't just dump jar on the path and call it a day.

**Comment by marypcbuk** [comment]
@majkinetor what a great way for the community to improve things ;)

**Comment by Witchilich** [comment]
Not just zip files. To support  all portable apps it should support 7z and cab files, and also things like youtube-dl. I dont know how you can implement yori in winget.
https://github.com/malxau/yori/issues/43
It is in Scoop
https://github.com/ScoopInstaller/Main/blob/master/bucket/yori.json
winget should also add shims for these apps and support powershell commands or scripts during installation and uninstallation
https://github.com/microsoft/winget-cli/issues/361
The aim should be portable apps support of the same level as Scoop.

**Comment by Witchilich** [comment]
I will also add, winget should be able extract installers from exes
https://download.microsoft.com/download/8/4/A/84A35BF1-DAFE-4AE8-82AF-AD2AE20B6B14/directx_Jun2010_redist.exe

**Comment by megamorf** [comment]
> I will also add, winget should be able extract installers from exes
> https://download.microsoft.com/download/8/4/A/84A35BF1-DAFE-4AE8-82AF-AD2AE20B6B14/directx_Jun2010_redist.exe

Why? Those typically have switches to pass parameters to the embedded msi files.

**Comment by Witchilich** [comment]
> > I will also add, winget should be able extract installers from exes
> > https://download.microsoft.com/download/8/4/A/84A35BF1-DAFE-4AE8-82AF-AD2AE20B6B14/directx_Jun2010_redist.exe
> 
> Why? Those typically have switches to pass parameters to the embedded msi files.
This one does not. You need to extract it with `.\directx_Jun2010_redist.exe /Q /T:C:\directx /C:/Silent` and then install through dxsetup.exe with `/silent`. winget has to extract it on its own, so that we can pass the silent argument.

**Comment by ckarcz** [comment]
instead of of supporting zip files (there are so many archive formats), why not have the manifest/winget have install script / uninstall scripts. that opens the door for zips, rar, 7z, fonts, or just arbitrary scripts that configure windows even.

**Comment by magicgoose** [comment]
> instead of of supporting zip files (there are so many archive formats), why not have the manifest/winget have install script / uninstall scripts. that opens the door for zips, rar, 7z, fonts, or just arbitrary scripts that configure windows even.

I agree that supporting other compression methods (zstd, lzma, etc.) and archive formats (7z, tar, etc.) is good and should eventually be added.  
Zips are the most common, so supporting just them is a good first step and I think it's better to have it ASAP and only after that think about other formats.  
Also I don't agree that it should complicate manifests in any way that requires scripting.  
Manifests should be as easy as possible to write and maintain, because that's going to be the primary factor affecting the quantity of available software.

**Comment by nu8** [comment]
https://github.com/microsoft/winget-pkgs/issues/2419

**Comment by jamesgecko** [comment]
> If we have a solution that works for everybody (Scoop), why trying to duplicate it?

This is a more recent thing, but Scoop development seems to be stalling out. The Scoop GCC package and all packages that depend on it have been broken for weeks. Chocolatey may soon be the only popular package manager that can install applications distributed via zip.

**Comment by motevets** [comment]
I noticed that this was assigned to a milestone that is now past due, but no one is assigned to it on Github. Is this being working on by someone at Microsoft? Is a pull request from the community welcome?

**Comment by denelon** [comment]
@motevets, we're just using milestones for sequencing. It's not being worked on yet. We're planning on tackling this with support for standalone .exe binaries at the same time. 

Part of the work will be to consider adding an ARP entry so these kinds of packages can be upgraded and removed. That would include adding and removing the appropriate path.

We'd be thrilled to have PRs from the community. The first part would be adding a feature toggle. Is this something you'd be interested in contributing to? I'd be happy to help work on a spec to capture the desired behavior and path as well as the changes to the schema and our validation mechanisms.

**Comment by motevets** [comment]
Yeah, I'd be interested in contributing. 😄

**Comment by JohnMcPMS** [comment]
This, and any other installer type where the actual file staging is handled by winget directly, will require the [`unvirtualizedResources` capability](https://docs.microsoft.com/en-us/windows/uwp/packaging/app-capability-declarations#restricted-capabilities) to be added to the MSIX package manifest.  In addition, we will need to use this to disable the file and registry virtualization via the referenced manifest elements.

File write virtualization needs to be disabled to allow the user to specify the location where the files will be placed.  It also gives us more flexibility to choose the default location.

Registry write virtualization needs to be disabled as we plan to write Add/Remove Program (ARP) entries for these.  This will allow a path for these packages to show up in the standard Windows flows for managing installed applications.  In addition, updating the user's stored path requires it (/speculation).

Note that the ability to disable virtualization was introduced in Windows 10.0.18362 (aka May 2019, 1903, or 19H1), and thus there will be cases where we do still have virtualization.  After some discussion, we have decided that this is acceptable for registry and we will need to write some code to ensure that files are not virtualized.  This will mean a limit on the locations that a pre-18362 client will be able to install to.

@motevets , I think you will probably want to apply the above capability and manifest elements to the dev package while you are working on it, otherwise there could be some nice long, confusing investigations on where your successful registry/file writes are ending up. 😃  We are also happy to accept any level of contribution, and I can understand not wanting to do the work to support the older OS.  Someone will need to do it before the feature can be enabled for everyone though.

**Comment by anionDev** [comment]
Thank you for this information JohnMcPMS, sounds interesting!

**Comment by JohnMcPMS** [comment]
We should also use https://github.com/ronomon/pure to check for malicious archives.  Depending on the performance, we could do it all of the time or only when the file came from an untrusted source / had a hash mismatch that was overridden.

**Comment by farag2** [comment]
When can we expect the zip supported release?

**Comment by denelon** [comment]
@farag2 we haven't got a date for .zip yet. It looks like it will be post 1.0. We've got quite a few Issues with "label:Issue-Bug" to squash first. There has been interest from a community member to contribute to that feature, so it might get done before we would otherwise get to it. It is one of the top asks along with #182 .exe and #221 PowerShell cmdlets.

**Comment by RichiCoder1** [comment]
Would it be feasible/reasonable to effectively take zips and then repackage them into msix/msixbundles using app execution aliases? It'd probably be heavy and have some catches I'm not thinking of, but would bring along the implicit benefits of msix and solve the shim issue (though that functionality is not without it's own issues).

**Comment by denelon** [comment]
@RichiCoder1 several packages have explicit distribution and licensing concerns that prevent us from automatically converting or modifying or distributing their software. We're looking at ways we can provide support publishers with respect to packaging their software installers via MSIX.

**Comment by denelon** [comment]
@c16a I'd be happy to discuss this with you. I'd expect @JohnMcPMS would also be interested. We did get the opportunity to meet with @motevets as well. Feel free to send an e-mail so we can try to coordinate a good time to chat. @brunoborges might also be interested with your use case for Java.

**Comment by c16a** [comment]
@denelon @brunoborges - I've written to you separately.

For everyone else, the following are the items that are being considered
1.	Use a path (like `%USERPROFILE%/WinGetApps`) as a package tree for all binary downloads.
2.	Supporting installation via .zip files
a.	Downloading files with checksum support.
b.	PowerShell scripts to process and add this to system PATH.
3.	Supporting installation of multiple versions of the same package
a.	Setting default versions across shells, and per-shell versions.

Of course, open to other suggestions in this area.

**Comment by RichiCoder1** [comment]
@c16a that all sounds awesome! Immediate feedback would be:

> Use a path (like %USERPROFILE%/WinGetApps) as a package tree for all binary downloads.

It'd be nice to support a System VS User-like switch here where you could either install to a well-known UserProfile directory or a well-known OS drive-level directory.

> PowerShell scripts to process and add this to system PATH.

It would be _really_ nice if PowerShell support was only necessary for complex edge-cases and simple zips/single file exes that are just xcopy & add to Path "Just Work".

**Comment by c16a** [comment]
> It'd be nice to support a System VS User-like switch here where you could either install to a well-known UserProfile directory or a well-known OS drive-level directory.

Yep, thanks for this. Might make sense for enterprise deployments too.

> It would be _really_ nice if PowerShell support was only necessary for complex edge-cases and simple zips/single file exes that are just xcopy & add to Path "Just Work".

I'm positive that it's straight forward for single file execs, but shell transformations might be required with zip extracts (stripping top level directories, renaming etc). Nevertheless, I get your point: when we speak about supported package types, `zip` and `exe` (say) could be treated separately.

**Comment by RichiCoder1** [comment]
> I'm positive that it's straight forward for single file execs, but shell transformations might be required with zip extracts (stripping top level directories, renaming etc). Nevertheless, I get your point: when we speak about supported package types, `zip` and `exe` (say) could be treated separately.

Yup. My biggest concern to add is just security/environment restrictions on running scripts. So ideally from both maintenance perspective and a security perspective, these common cases case work with some configuration and no script/shell involved unless absolutely necessary.

**Comment by c16a** [comment]
Yes, you're rightfully paranoid. Tools like **kubectl**, and **terraform** can straight up be moved into the install location in their own directories.

Here's my take on archives.

## Problems
1. What if the archive cannot be extracted by the Expand-Archive cmdlet (like GZ archives)? Supporting other archive formats currently requires another cmdlet to be installed, which is currently being worked on #163 
2. Some of them have top level directories, and some don't.

## Possible Solutions
### Checksums 
For the archive, and the PowerShell script (if any)

### Ensuring unauthorized file system access
We need to ensure that the above PS script does not make any file system changes beyond the scope of the installed directory for this version. Besides manual review, the only secure alternative that comes to mind is if the winget-cli has auditing control of the files being access by the Powershell script itself. 

Sorry for the Linux-speak, but like `strace` at process level. I'm assuming the PowerShell script would be run as a child process of the CLI, so we could monitor all file system calls that it makes. This can be used during integration testing. If Windows lets us monitor _and_ block all file system calls made by the current or child thread that don't pass some criteria, then the CLI can use this as an additional layer of security during installation as well.

Thoughts?

**Comment by RichiCoder1** [comment]
> What if the archive cannot be extracted by the Expand-Archive cmdlet (like GZ archives)? Supporting other archive formats currently requires another cmdlet to be installed, which is currently being worked on #163

To be honest, I assumed that archive & extraction support would be baked into WinGet itself as part of this. By extension, it'd become WinGet's responsibility to include support for a reasonable set of archive formats (.zip, .7z, .tar, .gz).

> Some of them have top level directories, and some don't.

With other managers, you'd just define this as part of the manifest.

Def agree that checksums are 100% required.

> Sorry for the Linux-speak, but like strace at process level. I'm assuming the PowerShell script would be run as a child process of the CLI, so we could monitor all file system calls that it makes. This can be used during integration testing. If Windows lets us monitor and block all file system calls made by the current or child thread that don't pass some criteria, then the CLI can use this as an additional layer of security during installation as well.

I know the capability to record PS calls exists (Transcripts) and I believe there's a way to lock down what CMDlets you can call, but I don't know about locking down FS calls specifically without getting knee deep in NTFS ACLs.

**Comment by c16a** [comment]
Thanks a lot for the inputs @RichiCoder1. I'll discuss them with the others and post timely updates.

**Comment by Alovchin91** [comment]
Just to support this suggestion:

This would make installing PowerShell 7 on Windows ARM64 much much easier 👍

**Comment by Xeeynamo** [comment]
Are there plans to create a PR for the ZIP support any time soon? I do not see any technical requirement draft anywhere.
If no one is currently working on it then I will be happy to start with a draft proposal on how I imagine WinGet could support compressed archives.

**Comment by denelon** [comment]
@Xeeynamo we're still catching up from the recent release activity and holiday. I'll see how soon I can pull a draft together for what we're thinking.

**Comment by Xeeynamo** [comment]
> @Xeeynamo we're still catching up from the recent release activity and holiday. I'll see how soon I can pull a draft together for what we're thinking.

Hi Denelon 👋thanks for the quick reply.

I have been thinking on how installing compressed archives (eg. ZIP) would work. Allow me to share my thoughts.

### The archive content is unpredictable

It can contains (but not limited to) one of the following options:

1) A stand-alone application. A single executable that is ready to be launched without performing any configuration (eg. installer).

2) Multiple applications. This is true for archives such as Sysinternals Suite, found on MS Docs.

3) An installer. If I recall correctly, downloadable Adobe products are a ZIP archive that contains an executable to run the installer. The installer itself can be an EXE, MSI or even a PS1 script.

As we already have a bunch of different scenarios, I would just focus on the first two as they might be the simplest implementation for a MVP:

### Make it User or Globally accessible

We want to have an output directory to extract our application(s) to. Based on a flag in the YML we might specify if it should be extracted in `%LOCALAPPDATA%` or, ideally, `%PROGRAMDATA%` based if we want to make the application globally available. It can be a string that can have either `user` or `global` as values.

### In-place upgrade?

We can easily agree that the output directory can be something like `%PROGRAMDATA%\publisher\package`, but this means that upgrades of the same application need to be in-place. This can lead to problems when a new version of the application removes a dependant file that was previously available, but the in-place upgrade would rather accumulate. Applications that rely on loading DLLs dynamically are affected by this problem.

An idea would be to clean-up the folder before performing the upgrade or just extract to `%PROGRAMDATA%\publisher\package\version`. This can be only done if we assume that the application would store the configuration in a different folder (eg. `%APPDATA%`), but as some applications store their configuration in the same extracted folder then it is easy to see how this cannot be viable.

There is also the problem of handling correctly the replacement or deletion of previous files if they are in use. Would winget just fail? What about rolling back the upgrade if it failed?

I could not find a way to solve this scenario yet and I do not believe that tracking the list of extracted files might be a clean solution either.

### Give an easily reachable entry-point

An installer **could** generate a desktop icon or a Start menu entry, but not an extracted archive. A quick win would be to have a list of paths where each one is a local path to the executable we want to add, as a shortcut, to the Start menu or the Desktop. There might also be the need to customise the icon the shortcut is going to use.

### A rough example

By collecting all my ideas, doubts and concerns I listed above, what I've been imaging could be an ideal YAML object is the following:

```yaml
Installers:
  - InstallerUrl: ""
    InstallerSha256: ""
    InstallerType: "zip"
    Scope: "user"
    Shortcuts:
      - Path: "procexp64.exe"
        Icon: "procexp64.exe,0"
        Startmenu: true
        Desktop: false
```

I know I am over-simplifying some other very specific scenario and that I might've been overlooking some critical aspects of this specific complex problem, but I hope this will bring some value to the issue to collectively reach a good draft for an MVP.
I hope I have not been too intrusive with my thoughts 🙏.

**Comment by brunoborges** [comment]
Hi @c16a,

I wanted to reply to your example regarding OpenJDK (Java). Today, the OpenJDK project is implemented in a way that it is smart enough to find where the JDK is installed. The binaries (e.g., `java`, `javac`, `jar`, and others) know where they are installed and can find the other files of the JDK. Simply put, the user only needs to point JAVA_HOME to a different JDK installation, and ensure %JAVA_HOME%/bin is part of %PATH%.

As you may know, Linux package managers don't implement the idea of 'switch/use/default'. This concept is always implemented by a different tool, designed oftentimes, for specific technologies. In Ubuntu/Debian, there is an `update-java-alternatives` command. In CentOS, there is an `update-alternative` command, and it is unclear if it supports JDK installations adequately.

In the Java world, there are better tools designed specifically to manage multiple JDKs installed in a local environment. I wrote about that [in this article](https://medium.com/@brunoborges/manage-multiple-jdks-on-mac-os-linux-and-windows-wsl2-3a73467b685c). 

All said I am not very much in favor of `winget` having such functionality, at the very least for managing JDKs, because on Windows, JDKs may be installed with MSI or simply have ZIP files extracted into somewhere. And different MSI installers for JDKs may have different settings/defaults, and I am certainly not in favor of winget forcing MSI or ZIP files for JDKs to conform to some standard. I think this would increase the scope of `winget` too much, and complicate things rather than simplify.

m2c

**Comment by denelon** [comment]
@Xeeynamo I wonder if there is a case to simplify the process of taking one of these .zip package and building an MSIX to handle the other concerns?  (Longer term possibility)

In the short term, feel free to build a draft PR for a spec. You can see we have a spec for the Dependency feature, and it is being broken down into bite sized parts for implementation. I would expect to follow a similar approach here so we can articulate what is proposed and gradually complete the implementation.

We would certainly want to start off with a Feature Toggle for .zip as an experimental feature. We would need to iterate on what the schema might look like and work through the right way for a .zip enabled client to handle this "new" manifest schema. Initially, we would only support "local" manifests with `winget install <package> -m <pathToManifest>` until we had a very solid set of capabilities. 

Making schema changes must cascade across the entire product ecosystem: 
*  Client
*  Validation Pipelines
*  Publishing Pipelines
*  REST API
*  wingetcreate

**Comment by yakatz** [comment]
This would be great to install packages like the Hugo static site generator too.

**Comment by edrohler** [comment]
Just wanted to mention that I am in the process of doing this with [Cmder](https://cmder.net). Not sure if it will help anyone wanted to get started now.

So far, I just created a wrapper [installer](https://github.com/edrohler/cmder-installer) repository for their zip distribution in NSIS.

I believe the process is similar to maintainers of packages in Linux.

Anyway, building an installer in NSIS using a zip file is pretty quick. To do this, I used the [NSIS Quick Setup Script Generator](https://nsis.sourceforge.io/NSIS_Quick_Setup_Script_Generator) to get started faster. The script just needed to be customized after initial generation.

Hope this helps anyone looking to get started now.

**Comment by lnlp** [comment]
> @denelon @brunoborges - I've written to you separately.
> 
> For everyone else, the following are the items that are being considered
> 
> 1. Use a path (like `%USERPROFILE%/WinGetApps`) as a package tree for all binary downloads.

No not %USERPROFILE% please, it's drive cannot be changed.
The C:\ drive is already becoming much bloated with non-system stuff.
Use %HOMEDRIVE%%HOMEPATH% instead which can be pointed to a different drive/folder.
Much better is to make such default location configurable per system and per user.

**Comment by pbek** [comment]
I wonder why `zip` is already used in [this testdata](https://github.com/microsoft/winget-cli/blob/master/src/AppInstallerCLITests/TestData/Manifest-Good-Spaces.yaml#L19).

**Comment by denelon** [comment]
@pbek We were expecting it to be a supported type very early on. Unfortunately, we had to adjust our expectations as we continued building the core functionality for 1.0. The service side took quite a bit more than estimated, and our current focus is stability, catalog health, and UX. We still plan to build support for .zip, but we have some other changes coming in the v1.1 manifest schema to support the more complex behaviors associated with "installing" a file downloaded as a .zip.

**Comment by denelon** [comment]
I didn't mean to close the issue. I hit the wrong button when UI jumped on me.

**Comment by pbek** [comment]
> We still plan to build support for .zip, but we have some other changes coming in the v1.1 manifest schema to support the more complex behaviors associated with "installing" a file downloaded as a .zip.

Thank you, @denelon.
What's about self extracting exe files? Are they also not supported for the same reasons?

**Comment by denelon** [comment]
@pbek, yes. It's likely we will have a similar implementation for both new types. We are considering building entries for them in Apps and Features so they can also be upgraded and installed. Ideally, publishers would build their own installers to handle these cases, but we wanted to make sure the other core functionality would work in either case. Sometimes a simple .exe will get the job done, but it's not really conducive to a package manager looking to manage software on a Windows system.

**Comment by pbek** [comment]
Too bad for apps like https://github.com/pbek/QOwnNotes that don't require installation. 😅

**Comment by denelon** [comment]
I think those could be supported. If we declare them as a portable/standalone .exe inside of a .zip, we would still capture the "version" and use the other metadata to create the appropriate entries in the system, add the proper PATH entry and it would be usable. Some of the work we're doing now for the 1.1 schema will make scenarios like this possible in the future.

**Comment by pbek** [comment]
That would only work for statically linked apps (like Go apps) with no other assets and not for dynamically linked apps like QOwnNotes (with all the Qt DLLs).

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
Surely other archive files such as tar should be allowed. Also, the supported methods of compression should be stated, or else no methods of compression should be allowed, else .zip, .tar, .zip.xz and .tar.gz (etcetera) shall also be vying for support.

**Comment by suusha13409486** [comment]
Thanks

**Comment by NicTanghe** [comment]
Here to request this and the fonts to be put in the 1.3 milestones.

**Comment by TheFanatr** [comment]
This is terrible. WinGet is seriously crippled without this, and getting this in an update really soon should be a priority. So many apps would really greatly benefit from WinGet distribution, but literally just can't even submit a package because of this one issue. Have had to literally just revert to community packages on Chocolatey, which are almost always out of date or doing some processing for days after release so for any zipped app I care about I just have to manage manually. Quite unfortunate and annoying.

**Comment by doctordns** [comment]
I would hope that the ability to handle zip files would get added at some future point. Is this feature request reflected in the road map ( I can't find it)? It should be.

**Comment by denelon** [comment]
@doctordns we have quite a few bugs to squash in the near term. 

We are tracking .zip #140 and portable applications #182 as [top 5 in the backlog](https://github.com/microsoft/winget-cli/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc) (along with native PowerShell #221 support) for new features on the client. I just don't yet know what the sequencing is going to be for all of them yet. We've been discussing whether to try and build an entry in Apps and Features so we can track "upgrade" and "remove" scenarios. Sadly there are a large number of edge cases to deal with.

**Comment by sredna** [comment]
> We've been discussing whether to try and build an entry in Apps and Features so we can track "upgrade" and "remove" scenarios.

What about naming the ARP **key** `Winget_%appname%`? On upgrade you simply update the DisplayName value inside this key. UninstallString would be something like `winget.exe --uninstallzip %appname%` and the InstallLocation value points to where the .zip was extracted.

**Comment by Suyashtnt** [comment]
probably a dumb question but why not use an external c++ library? I'm not a c++ dev but I would assume that they exist

**Comment by Rolling2405** [comment]
> What would we want to happen here? specify in the yaml where we want the folder to be unzipped to? Or ask the user where to unzip to?

I would want WinGet to automatically unzip the folder and install the program

**Comment by brunoborges** [comment]
So where goes the directory structure definition? Should all ZIP files follow the same structure?

**Comment by yakatz** [comment]
> What does it mean "install the program" once it is extracted from the ZIP file though?
> 
> I don't see neither yum or apt supporting installation of ZIP.

Technically, `deb` and `rpm` packages are archives (for `deb`, `tar`, and for `rpm`, `cpio`, both then `gzip`ed not `zip`) and the simplest installation just extracts the files to the locations specified by full paths or metadata. Most existing zip files couldn't be installed that way, but it is technically correct. The equivalent could be added to the manifest along with the zip file.

**Comment by brunoborges** [comment]
> Technically, `deb` and `rpm` packages are archives (for `deb`, `tar`, and for `rpm`, `cpio`, both then `gzip`ed not `zip`) and the simplest installation just extracts the files to the locations specified by full paths or metadata. Most existing zip files couldn't be installed that way, but it is technically correct. The equivalent could be added to the manifest along with the zip file.

Technically, `deb` and `rpm` are similar to `msi`.

Both DEB and RPM come with a spec just like MSI also has a spec. Of course all of them are compressed files.

But "supporting ZIP" would mean "supporting any compressed file without following a spec".

**Comment by brunoborges** [comment]
"just analyzing" certainly isn't simple, *especially* considering 20+ years of software that never bothered to produce MSI packages :-D

**Comment by yakatz** [comment]
What about extending the installer manifest (or having a zip manifest) that is added to the repository and contains this information?

**Comment by doctordns** [comment]
Seems to me that a zip file could contain at least two different things:
* the app itself - winget needs to extract it to some folder, and optionally create a shorcut to the main app.
*  an app installer application - winget needs to run extract the contents, then run the identified exe to do the installation.

Can the schema support both?

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
Why not have the repository handle the extraction?

**Comment by brunoborges** [comment]
@SeanFeldman no. `sysinternals` is pulled from MS Store. 

```powershell
> winget show sysinternals

Found Sysinternals Suite [9P7KNL5RWT25]
Version: Unknown
Publisher: Microsoft Corporation
Publisher Url: https://www.sysinternals.com
Description: Sysinternals Suite is a bundle of the Sysinternals utilities including Process Explorer, Process Monitor, Sysmon, Autoruns, ProcDump, all of the PsTools, and many more.

The Sysinternals website was created in 1996 by Mark Russinovich to host his advanced system utilities and technical information. Whether you’re an IT pro or a developer, you’ll find Sysinternals utilities to help you manage, troubleshoot, and diagnose your Windows systems and applications.

The suite includes AccessChk, AccessEnum, ADExplorer, ADInsight, ADRestore, Autologon, Autoruns, BGInfo, CacheSet, ClockRes, Contig, Coreinfo, CPU Stress, DebugView, Desktops, Disk2vhd, DiskExt, DiskMon, DiskView, DU (Disk Usage), EFSDump, FindLinks, Handle, Hex2dec, Junction, ListDLLs, LiveKd, LoadOrder, LogonSessions, MoveFile, NotMyFault, NTFSInfo, PendMoves, PipeList, ProcDump, Process Explorer, Process Monitor, PsExec, PsFile, PsGetSid, PsInfo, PsKill, PsList, PsLoggedOn, PsLogList, PsPasswd, PsPing, PsService, PsShutdown, PsSuspend, RamMap, RDCMan, RegDelNull, Regjump, RU (Registry Usage), SDelete, ShareEnum, ShellRunas, Sigcheck, Streams, Strings, Sync, Sysmon, TCPView, Testlimit, VMMap, VolumeID, WhoIs, WinObj, and ZoomIt.
Licence: https://live.sysinternals.com/Eula.txt
Privacy Url: https://privacy.microsoft.com/privacystatement
Agreements:
Category: Utilities & tools
Pricing: Free
Free Trial: No
Terms of Transaction: https://aka.ms/microsoft-store-terms-of-transaction
Seizure Warning: https://aka.ms/microsoft-store-seizure-warning
Store License Terms: https://aka.ms/microsoft-store-license

Installer:
  Type: MSStore
  Store Product Id: 9P7KNL5RWT25

**Comment by doctordns** [comment]
I had assumed 'zip file support'  meant support for .zip files.  Expanding support to cover additional files might be nice, but in the Windows world, those other formats are far less common. I cannot recall any 'compressed ' apps for Windows using anything other than .Zip. I would suggest doing .zip first, then look at customer needs.

**Comment by denelon** [comment]
We've been taking a "crawl", "walk", "run" approach when developing this product. There are so many edge cases, we wouldn't get many features in otherwise. For the current milestone, our primary focus is addressing bugs and improving the UX with our current set of features. I'm hoping we're able to start working on this and standalone executables for 1.3, but it's going to depend on how much other work we need to do on validation, publishing, and the future [integration with Intune](https://techcommunity.microsoft.com/t5/windows-it-pro-blog/evolving-the-microsoft-store-for-business-and-education/ba-p/2569423).

**Comment by Masamune3210** [comment]
7z supports RAR through unrar, so other than license issues surrounding that if any 7z could just be used since it also supports pretty much every other archive that would be even slightly common

**Comment by caiohamamura** [comment]
We need this feature! Basic web development tools are distributed as `.zip` such as `nginx` and `php`. Allowing to run custom `ps1` or `bat` scripts would be useful too. To install `composer` in `php`, for example, would also require running custom `php` scripts which could be called from a `ps1` or a `bat`.

For now I'm sticking with `chocolatey`, but it would be great to have a "native" tool, since it is already packed with current Windows 11.

**Comment by CharlotteEm** [comment]
Yes, this feature is strongly needed! But winget was specifically engineered to encourage users to depend on ancient proprietary MSI technology. Besides that, it is also loaded with privacy invasive telemetry.

**Comment by Masamune3210** [comment]
Then don't use it? If you disagree so much with how everything is being handled, just ignore the existence. You don't have to use it if you don't want to

**Comment by midoriiro** [comment]
Just added draft specification #1717

**Comment by dscalzi** [comment]
Might have been mentioned already, but this would be great for installing the Dart SDK (https://github.com/dart-lang/sdk/issues/43977)

**Comment by BentramLawrence** [comment]
everyone gets an email: https://github.com/microsoft/vscode/issues/138080

**Comment by bjoernf73** [comment]
This discussion certainly made my decision on stearing away from winget.

**Comment by bjoernf73** [comment]
> Masamune3210

Extreme disagreement to @Masamune3210 comments on this post. If you don't understand, don't comment.

**Comment by bjoernf73** [comment]
> We need this feature! Basic web development tools are distributed as `.zip` such as `nginx` and `php`. Allowing to run custom `ps1` or `bat` scripts would be useful too. To install `composer` in `php`, for example, would also require running custom `php` scripts which could be called from a `ps1` or a `bat`.
> 
> For now I'm sticking with `chocolatey`, but it would be great to have a "native" tool, since it is already packed within Windows 11.

I totally agree - chocolatey is way greater than winget. But it is a bit sad that such a great idea is torn down by people who doesn't want to relate to reality.

**Comment by midoriiro** [comment]
> > We need this feature! Basic web development tools are distributed as `.zip` such as `nginx` and `php`. Allowing to run custom `ps1` or `bat` scripts would be useful too. To install `composer` in `php`, for example, would also require running custom `php` scripts which could be called from a `ps1` or a `bat`.
> > For now I'm sticking with `chocolatey`, but it would be great to have a "native" tool, since it is already packed within Windows 11.
> 
> I totally agree - chocolatey is way greater than winget. But it is a bit sad that such a great idea is torn down by people who doesn't want to relate to reality.

What is the purpose of a ps1 or bat script. A post install script execution ?

**Comment by denelon** [comment]
I believe @Masamune3210 was responding to a post from an account that was deleted due to TOS violations.

**Comment by denelon** [comment]
We will be building support for .zip files.

I don't expect we will be supporting any scripting, however, as that is one of our original design goals.

**Comment by ajeetdsouza** [comment]
@denelon just to clarify -- support for zip files basically means that WinGet will "install" portable .exe files that are packaged inside a zip file, right? Would the executables be available in PATH after installation? This would be very handy to install CLI tools.

**Comment by denelon** [comment]
Initially, it may just be zipped installers, but we also have support for portable / standalone applications as a supported type planned as well. Once that's implemented, then yes, the scenario you are describing including the path will work.

**Comment by bjoernf73** [comment]
> @WtfJoke what's confusing? This is basically how most choco packages do it. They have a chocolateyInstall.ps1 that downloads content from the vendor and then runs the installation commands. Their nuspec file doesn't typically have the actual content because of legal redistribution rights issues.

Spot on!

**Comment by Masamune3210** [comment]
Yeah, was responding to someone who is no longer here. I mean no ill will to people who don't deserve it by showing just how little they think of others

**Comment by bjoernf73** [comment]
> > > We need this feature! Basic web development tools are distributed as `.zip` such as `nginx` and `php`. Allowing to run custom `ps1` or `bat` scripts would be useful too. To install `composer` in `php`, for example, would also require running custom `php` scripts which could be called from a `ps1` or a `bat`.
> > > For now I'm sticking with `chocolatey`, but it would be great to have a "native" tool, since it is already packed within Windows 11.
> > 
> > 
> > I totally agree - chocolatey is way greater than winget. But it is a bit sad that such a great idea is torn down by people who doesn't want to relate to reality.
> 
> What is the purpose of a ps1 or bat script. A post install script execution ?

The purpose is that you wont manage to cower the stuff you set out to do. In 2020, someone started this thread, telling people like yourself, that programs are simetimes distributed in a zip. There might be an installer in that zip, or just the binaries, that must be unzipped to a path in some %path%. That need is fearly simple to understand. Yet here we are. 2022 is coming, and you still feel the need to discuss it. That's why you need script abilities - to make winget an option for package management. Simply because developers like you understand reality slowly. The ball is in your court.

**Comment by WtfJoke** [comment]
> > @WtfJoke what's confusing? This is basically how most choco packages do it. They have a chocolateyInstall.ps1 that downloads content from the vendor and then runs the installation commands. Their nuspec file doesn't typically have the actual content because of legal redistribution rights issues.
> 
> Spot on!

Dont mean to sound rude, but I dont think your comment, which is referencing to a comment I made 1.5 year (Mai 2020) ago is contributing something to this issue (Neither does this one). 

I've explained myself what I didnt agree on (I didnt agree to creating custom unofficials installers to available official software which has ready to run executables insides a .zip).

**Comment by bjoernf73** [comment]
> > > @WtfJoke what's confusing? This is basically how most choco packages do it. They have a chocolateyInstall.ps1 that downloads content from the vendor and then runs the installation commands. Their nuspec file doesn't typically have the actual content because of legal redistribution rights issues.
> > 
> > 
> > Spot on!
> 
> Dont mean to sound rude, but I dont think your comment, which is referencing to a comment I made 1.5 year (Mai 2020) ago is contributing something to this issue (Neither does this one).
> 
> I've explained myself what I didnt agree on (I didnt agree to creating custom unofficials installers to available official software which has ready to run executables insides a .zip).

I was supporting you comment by saying that your comment is spot on.

**Comment by WtfJoke** [comment]
I see, than we had a classical missunderstanding, sorry about that 🙇 ☮️

**Comment by ajeetdsouza** [comment]
Can we try to keep this discussion polite and productive? All of us are waiting for this feature to be implemented, but insulting the maintainers is not going to help land that feature any faster. In fact, the maintainers are wasting even more time deleting your comments.

A lot of us are following this thread because we'd like to be notified when this eventually lands, and we end up getting an email every time somebody decides to be toxic because the feature is taking longer to land than they expected. Nobody is forcing you to use WinGet. If you feel that Scoop/Chocolatey are better package managers, then go ahead and use them. Please stop spamming this thread.

Another request -- since the maintainers have already agreed to work on this feature, comments like "this would be useful to implement X" are now irrelevant. It's sending a mail to everyone following this issue, for no good reason. Use the +1 instead, and trust the maintainers to prioritize features accordingly.

**Comment by berkeleynerd** [comment]
Looking forward to access to Hashicorp applications which appear to rely on this feature.

**Comment by midoriiro** [comment]
> > > > We need this feature! Basic web development tools are distributed as `.zip` such as `nginx` and `php`. Allowing to run custom `ps1` or `bat` scripts would be useful too. To install `composer` in `php`, for example, would also require running custom `php` scripts which could be called from a `ps1` or a `bat`.
> > > > For now I'm sticking with `chocolatey`, but it would be great to have a "native" tool, since it is already packed within Windows 11.
> > > 
> > > 
> > > I totally agree - chocolatey is way greater than winget. But it is a bit sad that such a great idea is torn down by people who doesn't want to relate to reality.
> > 
> > 
> > What is the purpose of a ps1 or bat script. A post install script execution ?
> 
> The purpose is that you wont manage to cower the stuff you set out to do. In 2020, someone started this thread, telling people like yourself, that programs are simetimes distributed in a zip. There might be an installer in that zip, or just the binaries, that must be unzipped to a path in some %path%. That need is fearly simple to understand. Yet here we are. 2022 is coming, and you still feel the need to discuss it. That's why you need script abilities - to make winget an option for package management. Simply because developers like you understand reality slowly. The ball is in your court.

I just asked a simple question to understand your point of view about this feature. You did not bother answer this simple question. So let me ask you again, why do you think this feature is needed for zip support feature ? Do you have a concrete example when post install script is mandatory ?

And yes, I feel the need to discuss it. Because that could be an interesting feature, not now, but later. Next time avoid insulting people, we are not on twitter :)

**Comment by doctordns** [comment]
If the team want to make an all-purpose package manager too for Windows, then having post-install actions (via a PowerShell script) would seem appropriate for the enterprise environment. Zip support too. Given the MS way of engineering, there should be some sort of provider model (i.e. not that different from the Secrets module) where developers can add support for new things (eg 7zip vs WinZip). 

What I am really missing is a good Vision and Scope document. At one time at MSFT, we used MSF (Microsoft Solutins Framework) to build things. MSF starts off with a clear Vision/Scope Document.  That document set out a clear picture of what is to be built along with the limitations of scope.

As I saw time and time again, in the absence of such a definitive document, potential users are assuming things., In this case, that a "package manager" does certain things. This results disappointed when what is delivered does not meet one's expectations,.

My view of a (Enterprise) package manager is the package management module in PowerShell. It does ALL I need at one level and (given it's open-source) could easily be extended to cover loading applications.  Instead, this team seems to have been trying to re-invent both the wheel, engine, ah more.

I come from a Server/Enterprise background - and for me, winget is little more than a developer power toy. I see no scope, today, for it in any Enterprise environment.  There are no PowerShell cmdlets, and the sSore is not supported (and probably never will be) on Windows Server. And so far, all the Winget PowerShell work seems to be about back end stuff. I doubt many organisations really care much about implementing REST APIs or private repos for internal use, for example.  Most IT Professionals just want to install a package (like they can use chocolatey) and move on and this means good cmdlets and ones that run on Server.

When I hear the term "Package Management for Windows" What I expect is a package manager that implements the following commands:

Find-Package - to find packages within the MS (or other) Package repo             
Get-Package - to find any package installed on a Windows system              
Install-Package - installs a packager whether from the MS/private repo, or from a locally saved copy
Save-Package  - downloads the package for installation using Install-Package
Uninstall-Package - to remove a package installed using Install-Package
Update-Package - to update packages installed using Install-Package.

We might usefully have a few more cmdlets to handle different repositories (Add-PackageRepository, Get-PackageRepositorry, and Remove-PackageRepository. 

If it were me running this project, I'd focus on making sure the community really agrees with the scope before writing a load of code no one will ever use. This thread suggests there is some difference of opinion over the scope and focus of the winget project that could be resolved once we had a better Vision//Scope document.

I would be happy to run an MSF course for the winget team for free to help! :-)

**Comment by bluehatgeek** [comment]
> Looking forward to access to Hashicorp applications which appear to rely on this feature.

I am looking for this option as well. There are a lot of tools that are coming as packaged zip files. Specifically a lot of DevOps tools but not only. It would be very handy to be able to add zip packages to the repository.

**Comment by sredna** [comment]
TL/DR: Use INF files to uninstall.

I recently started my own "package manager" as a side project and I started with .zip file support. Dealing with uninstall for the portable .zip packages is a bit of a pita. I did not want uninstallation to depend on my package manager still being on the system so my first PoC set the UninstallString to a `cmd.exe /k ...` command. This works but is rather ugly as the UI where you ask for confirmation will require the use of `choice` or `set /P`. QuietUninstallString with `cmd.exe /c ...` will briefly flash a console window. You might also at some point run into command line length limitations. Paths with `&` or `(` in them might be problematic.

Then I explored the possibility of storing the file list in the registry uninstall entry or even a full script in there but getting it back out with `reg query` or `reg export` proved to be problematic.

I finally settled on using a .inf file. This gives me support for:
 - Silent and a basic Yes/No native MessageBox confirmation "UI".
 - Require admin rights for all-users uninstalls.
 - Delete files relative to the location of the .inf.
 - Delete auto-generated Start menu entry.
 - Delete apps registry data (HKCU\Software\Sysinternals\ZoomIt etc.)
 - Zero external dependencies and will work on WinXP and later (the basic version should even work on Win95). 

The downside is that .inf files are a bit obscure and I'm already close to the limit of what is possible, not a whole lot of further capabilities can be added. You also end up with an extra file, ZoomIt_Uninst.inf etc. next to the extracted files.

I don't know what wingets plans are for uninstalling .zip file but I thought I'd at least share my findings.

**Comment by denelon** [comment]
Feel free to review the [specification PR](https://github.com/microsoft/winget-cli/pull/2270).

**Comment by KrisSchneider** [comment]
I have an idea. I think adding 7zip CLI to winget would solve the problem of winget-cli not being able to extract archive files.

Here's a link to 7zip CLI:  https://www.7-zip.org/download.html 

With 7z CLI, you can extract an archive's contents into a new folder with a command like this.

Example:
```powershell 
7z x "C:\Users\krist\Downloads\AwesomeProgram.zip" -o"$home/Downloads/*" -bb3
```


What do you all think? Would this cause any issues with winget-cli, or would it work just fine?

**Comment by KrisSchneider** [comment]
This is why I think using **7zip CLI** would be a much better choice in comparison to `Expand-Archive`.

Here is a demonstration:
![(WindowsTerminal)s8g4V6_07-20-2022](https://user-images.githubusercontent.com/101912712/179979457-c2af1f2a-5769-4db4-8fb4-5dd83ecc40ed.jpg)

```powershell
PS C:\Users\krist> 7z x "C:\Users\krist\Downloads\Ex.zip" -o"$home\Downloads\Ex2"

7-Zip 22.01 (x64) : Copyright (c) 1999-2022 Igor Pavlov : 2022-07-15

Scanning the drive for archives:
1 file, 1387152673 bytes (1323 MiB)

Extracting archive: C:\Users\krist\Downloads\Ex.zip
--
Path = C:\Users\krist\Downloads\Ex.zip
Type = zip
Physical Size = 1387152673

Everything is Ok

Files: 8
Size:       1388931888
Compressed: 1387152673
PS C:\Users\krist> 7z x "C:\Users\krist\Downloads\Example.zip" -o"$home\Downloads\Ex1"

7-Zip 22.01 (x64) : Copyright (c) 1999-2022 Igor Pavlov : 2022-07-15

Scanning the drive for archives:
1 file, 1380993596 bytes (1318 MiB)

Extracting archive: C:\Users\krist\Downloads\Example.zip
--
Path = C:\Users\krist\Downloads\Example.zip
Type = zip
Physical Size = 1380993596

Everything is Ok

Files: 8
Size:       1388931888
Compressed: 1380993596
PS C:\Users\krist> 7z x "C:\Users\krist\Downloads\AwesomeProgram.7z" -o"$home\Downloads\Ex3"

7-Zip 22.01 (x64) : Copyright (c) 1999-2022 Igor Pavlov : 2022-07-15

Scanning the drive for archives:
1 file, 15013876 bytes (15 MiB)

Extracting archive: C:\Users\krist\Downloads\AwesomeProgram.7z
--
Path = C:\Users\krist\Downloads\AwesomeProgram.7z
Type = 7z
Physical Size = 15013876
Headers Size = 2306
Method = LZMA2:26 LZMA:20 BCJ2
Solid = +
Blocks = 2

Everything is Ok

Folders: 25
Files: 131
Size:       66826375
Compressed: 15013876
PS C:\Users\krist>
```

As you can see, the `7z` command from **7zip CLI** has no issues extracting all the archives into an individual new folder compared to the `Expand-Archive` command.

**Comment by doctordns** [comment]
I can see that 7Zip may have benefits. But at the same time, I would imagine the dependency on a third-party, non-MSFT product. The legal implications alone are most interesting (and I am NOT a lawyer, but have seen MS Legal in action). I also suspect that applications only issued in 7zip own compression format is relatively low.  

Does anyone have statistics on how many applications there are that Winget could NOT install without 7ZIP support? If the numbers are not very high,  I can understand why this suggestion might get rejected.  

Of course, since (I think) the product is open source, there would be little to stand in the way of the community adding 7Zip native support. I hope the product is sufficiently layered to enable another compression mechanism to be inserted into the stack. I suspect that that will not be as easy as it could be. Nevertheless, there is (I think) always that opportunity if there IS a huge demand.

my .02€ worth.

**Comment by KrisSchneider** [comment]
> But at the same time, I would imagine the dependency on a third-party, non-MSFT product. The legal implications alone are most interesting (and I am NOT a lawyer, but have seen MS Legal in action).

@doctordns Does Microsoft not approve of the third-party software in their products? Also, the developer **Igor Pavlov** is fine with the use of his software in other products. As long as he gets credited as being the developer of 7Zip.

There is a license page in the 7zip website that shows what you are allowed to do, and what you are not allowed with 7Zip.
https://www.7-zip.org/license.txt

**Comment by RichiCoder1** [comment]
> There is a problem to that. What if the developer/publisher or a redistributor chooses to compress a folder/files into a .zip file using LZMA(or any other compression) method instead of using the Deflate method?

Worth calling out that any publisher publishing to Windows will default to Zip/Deflate by default. It's ubiquitous and simple and built-in to the explorer. Not doing this is a highly exception case and only done if they have their own installation method.

> Does Microsoft not approve of the third-party software in their products? Also, the developer Igor Pavlov is fine with the use of his software in other products. As long as he gets credited as being the developer of 7Zip.

Worth noting that in order to support 7z, Microsoft would not only have to take on legal accreditation, but bundling, validation against all winget's targets, security, maintenance, and a number of other issues which get more complicated with a third party component. Something that could be explored post-mvp, but pose significant blockers for getting support out the door.

There _is_ something that other package managers do where once #163 lands, a package could take a dependency on 7z and mark themselves as being unpacked by it which would another way, but that introduces it's own complexities.

**Comment by denelon** [comment]
We've been discussing what it might look like to support other compression algorithms.
* .gz
* .tar
* .tar.gz

That does introduce complexity to the manifest. As @RichiCoder1 pointed out, there are many other long-term implications. We want to get something out the door, and then we can look at the value of the edge cases. 

I would also like some data points like @doctordns asked. I don't know how many packages use algorithms that wouldn't be supported. That would be a factor in determining the value and relative priority of looking to support other formats.

**Comment by KrisSchneider** [comment]
> Worth noting that in order to support 7z, Microsoft would not only have to take on legal accreditation, but bundling, validation against all winget's targets, security, maintenance, and a number of other issues which get more complicated with a third party component. Something that could be explored post-mvp, but pose significant blockers for getting support out the door.
> 
> There is something that other package managers do where once https://github.com/microsoft/winget-cli/issues/163 lands, a package could take a dependency on 7z and mark themselves as being unpacked by it which would another way, but that introduces it's own complexities.

Oh, so that's why. Adding a third party component would make winget more complicated to manage, in comparison to using a native tool already built into Windows File Explorer. I hope I didn't misunderstand this take.

**Comment by KrisSchneider** [comment]
> I would also like some data points like @doctordns asked. I don't know how many packages use algorithms that wouldn't be supported. That would be a factor in determining the value and relative priority of looking to support other formats.

So, when winget eventually supports .zip archives in the future. Other archive formats will also be supported, depending on their relative value and priority. 

That's good to hear, I hope you winget devs are blessed with luck in overcoming the difficulties that this feature development might bring.

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
Support for 7-zip should be unnecessary if Extract-Archive continues to improve as "http://github.com/dotnet/runtime/issues/65951" demonstrates.

**Comment by doctordns** [comment]
@KrisSchneider you got it right. Taking a dependency on a third-party product represents a long-term maintenance and distribution commitment. It would require a lot of validation of their code, plus it bloats the installation package and makes the overall code more complex than might otherwise be the case. It is not quick, easy, or cheap. In my experience, it is not something any MSFT team takes on lightly. That is not to say that the team should NOT do 7zip. But only if there is demonstrable customers need.  

To justify a serious consideration, the team would need evidence.  Just how many packages are there that Winget (with Zip support) could not handle but would be ok IF there was 7zip support? My feeling today is that the number would be very, very low. But I could be wrong. 

Absent such evidence, I think this issue could be resolved as Will Not Fix. But with the general understanding that when the evidence IS produced, the team can re-examine based on demonstrated customer needs. 

At the same time, I would hope that the team layers ZIP support in such a way as to make it straightforward to add 7zip support without major work to Winget.  I have always been impressed with the elegance that many MSFT products had with layering (eg, network stack mini drivers, etc.).

**Comment by RichiCoder1** [comment]
> @KrisSchneider and @RichiCoder1, support for `7zip` should be unnecessary, because "[http://github.com/PowerShell/Microsoft.PowerShell.Archive/issues/22](https://github.com/PowerShell/Microsoft.PowerShell.Archive/issues/22)" already requests support for the relevant format[s], and addition of the necessary functionality, when "[http://github.com/dotnet/runtime/issues/65951](https://github.com/dotnet/runtime/issues/65951)" has been completed, should be consequently easy.

WinGet does not use, nor have a dependency on either .NET or PowerShell Core. I

> Absent such evidence, I think this issue could be resolved as Will Not Fix. But with the general understanding that when the evidence IS produced, the team can re-examine based on demonstrated customer needs.

(Worth noting this ticket is about the `.zip` support that's currently in progress, and 7zip has no relation but is a feature ask in addition to base `.zip` support).

**Comment by reyqn** [comment]
Maybe I'm wrong, but there has been lots of pings for lots of people subscribed to this issue, even though kind of everything was already talked about in this issue, and the specs for zip install (and particularly extraction) were already drafted [here.](https://github.com/microsoft/winget-cli/pull/2270)

I get that people want to participate, but please be wary that lots of other people receive notifications at every comment on those kind of long awaited issues. To those I say sorry for the ping.

**Comment by IanDarwin** [comment]
> +1 for adding the ability to execute a `.ps1` or `.bat` script after extracting a ZIP file. Ideally, this should be a file that may not be included in the original package. This should allow us to work with most types of packages distributed as ZIP files.

-1 for running semi-arbitrary scripts during a privileged install. Security? What was that?

**Comment by Masamune3210** [comment]
If it was being included with the manifest I would assume some validation of actions the script performed would be in place or at least in planning

**Comment by Masamune3210** [comment]
There is already a VM used to validate that the install is performed correctly, so it's probably not super outlandish to extend that to validating scripts as well

**Comment by rokejulianlockhart** [comment]
Should https://github.com/microsoft/winget-cli-restsource/issues/117 be closed?

**Comment by rollingmoai** [comment]
Winget has to support other archive formats as well. See	https://github.com/microsoft/winget-cli/issues/2899
